### ISSUE TITLE
Switch between header/source finds incorrect files in trivial cases.

### ISSUE THREAD

**Issue by akb825** [clangd/clangd#1617]
When header and source files are in separate folders, swapping between header and source files fail in trivial situations:
1. .h and .cpp files with the same name but different folders will fail to swap if the contents are empty.
2. Similar to the above point, if there are build errors it may fail to execute the swap.
3. If a header only contains forward declarations and functions (no class definition), it will go to the file that contains the first class declaration.

For an example of point 3, see [test-switch.tar.gz](https://github.com/clangd/clangd/files/11378790/test-switch.tar.gz). This was tested with Qt Creator 10.0.0, which currently uses the clangd code model to handle header/source switches.

1. Open Foo.cpp.
2. Execute Switch Header/Source, it will correctly switch to Foo.h.
3. Execute Switch Header/Source again from Foo.h, notice how instead of switching to Foo.h it switches to Bar.h as the "source" file.
4. If you delete the `class Bar;` forward declare and change the function declaration to `void foo(class Bar& bar);`, executing Switch Header/Source correctly switches to Foo.cpp.

**Logs**

[log.txt](https://github.com/clangd/clangd/files/11378784/log.txt)

**System information**

Output of `clangd --version`: 
clangd version 16.0.2 (https://github.com/llvm/llvm-project 18ddebe1a1a9bde349441631365f0472e9693520)
Features: linux+grpc
Platform: x86_64-unknown-linux-gnu

Editor/LSP plugin: Qt Creator 10.0.0

Operating system: Arch Linux

**Comment by HighCommander4** [comment]
See https://github.com/clangd/clangd/issues/1043 for some discussion of how this feature currently works.

I think the "file index" approach discussed in [this comment](https://github.com/clangd/clangd/issues/1043#issuecomment-1062946988) is a promising one and should address the case of files with the same name in different directories, regardless of their contents.

There are probably opportunities to improve the content-based heuristics as well, but the above sounds like the bigger bang for buck overall.

**Comment by akb825** [comment]
I agree, I think a file index and using heuristics based on that would handle situations like this much better overall. Care would need to be taken if there's multiple files with the same name, though, ideally ties would be broken based on more similar paths. (e.g. if multiple internal projects with a Foo.h and Foo.cpp pair, one heuristic could be to match the companion that has the fewest directories to a common root)

I saw that issue earlier, but it looks like their fix only resolved issues for differing case for case-insensitive filesystems.

**Comment by MartyMcFlyInTheSky** [comment]
I want to throw in the case where we have multiple header files per definition (and maybe vice versa). The switch should really be a cycle that you work through on each command invocation. Compare [this](https://github.com/clangd/clangd/issues/1898) and [this](https://github.com/clangd/clangd/issues/1797) issue.

Idea:
1) Generate a graph of all the files involved, locate all headers and corresponding implementations and vice versa.
2) Arrange them statistically, so that you would first jump to the file that maps most of the declarations / definitions with regards to the current file. This is usually what you want. Then for each further jump you cycle through the less 'relevant' files until you arrive at your original file.
